File: client/web/src/lib/api-client.ts
import { useAuthStore } from "@/stores/useAuthStore";
import { apiBaseUrl } from "@/lib/base-path";

const API_BASE_URL = import.meta.env.VITE_API_BASE_URL ?? apiBaseUrl;

export class ApiError extends Error {
  public status: number;

  constructor(status: number, message: string) {
    super(message);
    this.name = "ApiError";
    this.status = status;
  }
}

/** Network-layer failure (server unreachable, DNS, TLS) — no HTTP status. */
export class NetworkError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "NetworkError";
  }
}

function createApiUrl(endpoint: string): string {
  return `${API_BASE_URL}${endpoint}`;
}

let refreshPromise: Promise<void> | null = null;

async function doFetch(
  endpoint: string,
  method: string,
  body?: unknown,
  headers?: Record<string, string>,
): Promise<Response> {
  try {
    return await fetch(createApiUrl(endpoint), {
      method,
      headers: {
        "Content-Type": "application/json",
        ...headers,
      },
      body: body ? JSON.stringify(body) : undefined,
      credentials: "include",
    });
  } catch {
    // fetch() throws TypeError("Failed to fetch") on DNS failure, connection
    // refused, CORS rejection, etc. The underlying cause isn't available.
    throw new NetworkError("Could not reach the server.");
  }
}

/**
 * Build a user-facing error message from a non-success response. Parses the
 * server's JSON `detail` field only when the body is actually JSON; otherwise
 * returns a status-code-based message. Never returns raw body text —
 * prevents reverse-proxy HTML error pages from leaking into the UI.
 */
async function errorMessage(response: Response): Promise<string> {
  const contentType = response.headers.get("content-type") ?? "";
  const body = await response.text();
  if (contentType.includes("json") || body.trimStart().startsWith("{")) {
    try {
      const parsed = JSON.parse(body) as { detail?: unknown };
      if (typeof parsed.detail === "string" && parsed.detail.trim()) {
        return parsed.detail;
      }
    } catch {
      // Fall through to status-code message.
    }
  }
  switch (response.status) {
    case 400: return "Bad request (HTTP 400)";
    case 403: return "Access denied (HTTP 403)";
    case 404: return "Not found (HTTP 404)";
    case 500: return "Server error (HTTP 500)";
    case 502:
    case 503: return `Service unavailable (HTTP ${response.status})`;
    case 504: return "Server did not respond in time (HTTP 504)";
    default: return `Server returned HTTP ${response.status}`;
  }
}

async function handleResponse<T>(response: Response): Promise<T> {
  if (!response.ok) {
    throw new ApiError(response.status, await errorMessage(response));
  }

  if (response.status === 204 || response.headers.get("content-length") === "0") {
    return undefined as T;
  }

  return response.json() as Promise<T>;
}

const AUTH_ENDPOINTS = ["/auth/login", "/auth/login/complete", "/auth/register", "/auth/refresh", "/auth/logout"];

function isAuthEndpoint(endpoint: string): boolean {
  return AUTH_ENDPOINTS.some((path) => endpoint.startsWith(path));
}

async function request<T>(
  endpoint: string,
  method: string,
  body?: unknown,
  headers?: Record<string, string>,
): Promise<T> {
  const response = await doFetch(endpoint, method, body, headers);

  if (response.status === 401 && !isAuthEndpoint(endpoint)) {
    // Coalesce concurrent refresh attempts into a single request
    if (!refreshPromise) {
      refreshPromise = doFetch("/auth/refresh", "POST", {})
        .then((r) => {
          if (!r.ok) throw new ApiError(r.status, "Token refresh failed");
        })
        .finally(() => {
          refreshPromise = null;
        });
    }

    try {
      await refreshPromise;
    } catch (err) {
      // A network failure during refresh doesn't mean the session expired —
      // the server was just unreachable. Keep the tokens so the user can
      // retry once connectivity returns, and surface the network error.
      if (err instanceof NetworkError) throw err;
      useAuthStore.getState().logout();
      throw new ApiError(401, "Session expired");
    }

    // Retry original request with fresh cookies
    const retryResponse = await doFetch(endpoint, method, body, headers);
    return handleResponse<T>(retryResponse);
  }

  return handleResponse<T>(response);
}

export const apiClient = {
  request,
  get: <T>(endpoint: string, headers?: Record<string, string>) =>
    request<T>(endpoint, "GET", undefined, headers),
  post: <T>(endpoint: string, body?: unknown, headers?: Record<string, string>) =>
    request<T>(endpoint, "POST", body, headers),
  put: <T>(endpoint: string, body?: unknown, headers?: Record<string, string>) =>
    request<T>(endpoint, "PUT", body, headers),
  delete: <T>(endpoint: string, headers?: Record<string, string>) =>
    request<T>(endpoint, "DELETE", undefined, headers),
  logout: async () => {
    try {
      await doFetch("/auth/logout", "POST", {});
    } catch {
      // Best-effort: clear client state even if server call fails
    }
  },

  /** POST JSON, return raw Response (for blob downloads). */
  postRaw: (endpoint: string, body?: unknown): Promise<Response> =>
    doFetch(endpoint, "POST", body),

  /** POST FormData (for file uploads). */
  postForm: async <T>(endpoint: string, formData: FormData): Promise<T> => {
    const response = await fetch(createApiUrl(endpoint), {
      method: "POST",
      credentials: "include",
      body: formData,
    });
    return handleResponse<T>(response);
  },
};

File: client/android/core/src/main/java/io/github/pstanar/pstotp/core/api/ApiClient.kt
package io.github.pstanar.pstotp.core.api

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import io.github.pstanar.pstotp.core.model.api.RefreshResponse
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response
import org.json.JSONObject
import java.io.IOException
import java.net.ConnectException
import java.net.SocketTimeoutException
import java.net.UnknownHostException
import java.util.concurrent.TimeUnit
import javax.net.ssl.SSLException

/**
 * HTTP client wrapping OkHttp with Bearer token management.
 *
 * - Injects Authorization header on authenticated requests
 * - On 401: acquires mutex, refreshes tokens, retries original request
 * - On refresh failure: throws SessionExpiredException
 */
class ApiClient(baseUrl: String) {

    /** Base URL without trailing slash (e.g., "https://totp.example.com/api"). */
    var baseUrl: String = baseUrl.trimEnd('/')

    var accessToken: String? = null
    var refreshToken: String? = null

    /** Called after token refresh so callers can persist new tokens to SecureStore. */
    var onTokensRefreshed: ((access: String, refresh: String) -> Unit)? = null

    private val refreshMutex = Mutex()

    private val client = OkHttpClient.Builder()
        .connectTimeout(15, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .writeTimeout(30, TimeUnit.SECONDS)
        .build()

    private val jsonMediaType = "application/json; charset=utf-8".toMediaType()

    fun setTokens(access: String?, refresh: String?) {
        accessToken = access
        refreshToken = refresh
    }

    fun clearTokens() {
        accessToken = null
        refreshToken = null
    }

    fun hasTokens(): Boolean = accessToken != null && refreshToken != null

    // --- HTTP methods ---

    suspend fun <T> get(endpoint: String, parser: (JSONObject) -> T): T =
        executeWithAuth("GET", endpoint, body = null, parser = parser)

    suspend fun <T> post(endpoint: String, body: JSONObject?, parser: (JSONObject) -> T): T =
        executeWithAuth("POST", endpoint, body, parser)

    suspend fun <T> put(endpoint: String, body: JSONObject?, parser: (JSONObject) -> T): T =
        executeWithAuth("PUT", endpoint, body, parser)

    suspend fun delete(endpoint: String) {
        executeWithAuth<Unit>("DELETE", endpoint, body = null) { }
    }

    /** POST without auth (for login challenge, register, etc.). */
    suspend fun <T> postPublic(endpoint: String, body: JSONObject?, parser: (JSONObject) -> T): T =
        execute("POST", endpoint, body, addAuth = false, parser = parser)

    // --- Internal ---

    private suspend fun <T> executeWithAuth(
        method: String,
        endpoint: String,
        body: JSONObject?,
        parser: (JSONObject) -> T,
    ): T {
        return try {
            execute(method, endpoint, body, addAuth = true, parser = parser)
        } catch (e: ApiException) {
            if (e.statusCode == 401) {
                refreshTokens()
                execute(method, endpoint, body, addAuth = true, parser = parser)
            } else {
                throw e
            }
        }
    }

    private suspend fun refreshTokens() {
        refreshMutex.withLock {
            val rt = refreshToken ?: throw SessionExpiredException()
            val refreshBody = JSONObject().put("refreshToken", rt)
            try {
                val response = execute("POST", "/auth/refresh", refreshBody, addAuth = false) {
                    RefreshResponse.fromJson(it)
                }
                accessToken = response.accessToken
                refreshToken = response.refreshToken
                if (response.accessToken != null && response.refreshToken != null) {
                    onTokensRefreshed?.invoke(response.accessToken, response.refreshToken)
                }
            } catch (e: NetworkException) {
                // Server unreachable — tokens may still be valid. Don't
                // destroy the session; let the caller surface a network
                // error and let the user retry once connectivity is back.
                throw e
            } catch (_: Exception) {
                clearTokens()
                throw SessionExpiredException()
            }
        }
    }

    private suspend fun <T> execute(
        method: String,
        endpoint: String,
        body: JSONObject?,
        addAuth: Boolean,
        parser: (JSONObject) -> T,
    ): T = withContext(Dispatchers.IO) {
        val url = "$baseUrl$endpoint"

        val requestBody = body?.toString()?.toRequestBody(jsonMediaType)

        val request = Request.Builder()
            .url(url)
            .apply {
                when (method) {
                    "GET" -> get()
                    "POST" -> post(requestBody ?: "".toRequestBody(jsonMediaType))
                    "PUT" -> put(requestBody ?: "".toRequestBody(jsonMediaType))
                    "DELETE" -> delete(requestBody)
                }
                if (addAuth) {
                    accessToken?.let { header("Authorization", "Bearer $it") }
                }
            }
            .build()

        val response = try {
            client.newCall(request).execute()
        } catch (e: IOException) {
            throw NetworkException(friendlyNetworkMessage(e), e)
        }
        val responseBody = response.body?.string()

        if (!response.isSuccessful) {
            val message = errorMessage(response, responseBody)
            when (response.code) {
                401 -> throw ApiException(401, message)
                409 -> throw ConflictException(message)
                else -> throw ApiException(response.code, message)
            }
        }

        if (responseBody.isNullOrBlank()) {
            parser(JSONObject())
        } else {
            parser(JSONObject(responseBody))
        }
    }

    private fun errorMessage(response: Response, body: String?): String =
        buildErrorMessage(response.code, response.body?.contentType()?.subtype, body)

    private fun friendlyNetworkMessage(e: IOException): String = when (e) {
        is UnknownHostException -> "Server not found. Check the URL and your connection."
        is SocketTimeoutException -> "Server did not respond in time."
        is ConnectException -> "Could not reach the server."
        is SSLException -> "Secure connection failed (TLS/certificate issue)."
        else -> e.message?.takeIf { it.isNotBlank() } ?: "Network error"
    }
}

/**
 * Build a user-facing error message for a non-success response. Parses the
 * JSON `detail` field when the body is JSON; otherwise produces a status-
 * code-based message. Never returns raw body text — prevents reverse-proxy
 * HTML error pages from leaking into the UI. Top-level so it's unit-testable
 * without spinning up OkHttp.
 */
internal fun buildErrorMessage(statusCode: Int, contentTypeSubtype: String?, body: String?): String {
    val looksLikeJson = contentTypeSubtype?.contains("json", ignoreCase = true) == true ||
        body?.trimStart()?.startsWith("{") == true
    if (looksLikeJson && !body.isNullOrBlank()) {
        runCatching {
            val detail = JSONObject(body).optString("detail").takeIf { it.isNotBlank() }
            if (detail != null) return detail
        }
    }
    return when (statusCode) {
        400 -> "Bad request (HTTP 400)"
        403 -> "Access denied (HTTP 403)"
        404 -> "Not found (HTTP 404)"
        500 -> "Server error (HTTP 500)"
        502, 503 -> "Service unavailable (HTTP $statusCode)"
        504 -> "Server did not respond in time (HTTP 504)"
        else -> "Server returned HTTP $statusCode"
    }
}
